uniffi_bindgen/src/bindings/python/templates/ObjectTemplate.py

```diff
@@ -14,17 +20,37 @@ def __del__(self):
             self._handle
         )
 
-    {% for cons in obj.alternate_constructors() -%}
+    # Used by alternative constructors or any methods which return this type.
+    # May return an existing instance, or create a new one if we can't find one.
     @classmethod
-    def {{ cons.name()|fn_name_py }}(cls, {% call py::arg_list_decl(cons) %}):
-        {%- call py::coerce_args_extra_indent(cons) %}
-        # Call the (fallible) function before creating any half-baked object instances.
-        handle = {% call py::to_ffi_call(cons) %}
+    def _get_or_make_instance_(cls, handle):
+        {# Look in our weak map for an instance already associated with this
+           pointer. If it exists we can return it, but the pointer we have
+           is a clone of an `Arc<>` so has a reference we need to destroy.
+        #}
+        existing = cls._instances.get(handle)
+        if existing is not None:
+            rust_call_with_error(
+                InternalError,
+                _UniFFILib.{{ obj.ffi_object_free().name() }},
+                handle
+            )
```

Comment:
Oh! I just understood what you're doing here, and it's neat.

In my head I was worried about the asymmetry where we return a fresh clone of the `Arc<T>` on the Rust side but collapse them all into a single instance on the foreign-language side. I thought we might have to keep track of the number of `Arc<T>` that had been merged into this instance, so that we could call `ffi_object_free` the appropriate number of times. But immediately destroying the clones is much simpler (and really reinforces the "The Prestige" metaphor that @jhugman used to describe this identity map idea to me at one point...).

In the future we might be motivated to keep a count and destroy them all at once for performance reasons, but that seems likely to be a long ways off.